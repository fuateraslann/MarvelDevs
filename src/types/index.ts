export type TImageExtensions = 'jpg' | 'png' | 'svg'
export type TCharacter = {
  id: number
  name: string
  description: string
  thumbnail: {
    path: string
    extension: TImageExtensions
  }
}
export type TComics = {
  id: number
  title: string
  thumbnail: {
    path: string
    extension: TImageExtensions
  }
}
