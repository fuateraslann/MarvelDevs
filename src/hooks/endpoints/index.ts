import { useQuery, UseQueryResult } from 'react-query'

import { axios } from 'utils'
import { TCharacter } from 'types'

export const useGetCharacters = (): UseQueryResult<TCharacter[]> => {
  return useQuery(
    'characters',
    async (): Promise<TCharacter[]> => {
      return await axios
        .get(`/v1/public/characters`)
        .then((response) => {
          return response.data.data.results
        })
        .catch((error) => {
          console.log(error)
        })
    },
    {
      refetchOnWindowFocus: false,
    }
  )
}
