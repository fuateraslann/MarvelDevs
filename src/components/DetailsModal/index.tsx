import {
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  Typography,
  Button,
  Grid,
} from '@mui/material'

import { useGetCharacterComics } from 'hooks/endpoints'
import { TCharacter } from 'types'
import ComicsList from '../ComicsList'

interface IDetailsModalProps {
  detailsItem: TCharacter
  onCloseModal: () => void
  openModal: boolean
}
const DetailsModal = ({
  detailsItem,
  onCloseModal,
  openModal,
}: IDetailsModalProps) => {
  const { data: comics, isLoading: isLoadingComics } = useGetCharacterComics({
    characterId: detailsItem.id,
    enabled: openModal,
  })

  const imageSrc =
    detailsItem.thumbnail.path + '.' + detailsItem.thumbnail.extension

  return (
    <Dialog open={openModal} onClose={onCloseModal}>
      <DialogTitle>
        <h4> Details Of {detailsItem.name}</h4>
      </DialogTitle>
      <DialogContent
        sx={{
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'space-between',
        }}
      >
        <Grid
          sx={{
            display: 'flex',
            flexDirection: 'row',
            justifyContent: 'space-between',
          }}
        >
          <Grid flex={2}>
            <img style={{ width: '90%', height: '90%' }} src={imageSrc} />
            <Typography variant="h6" gutterBottom>
              {detailsItem.name}
            </Typography>
          </Grid>
          <Grid flex={1}>
            {detailsItem.description !== ''
              ? detailsItem.description
              : 'No Description'}
          </Grid>
        </Grid>
        <Grid flex={1}>
          <h4>Comics Of {detailsItem.name}</h4>
          <ComicsList comics={comics ?? []} isLoading={isLoadingComics} />
        </Grid>
      </DialogContent>
      <DialogActions>
        <Button onClick={onCloseModal}>Close</Button>
      </DialogActions>
    </Dialog>
  )
}

export default DetailsModal
