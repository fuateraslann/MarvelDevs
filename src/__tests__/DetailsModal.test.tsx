import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import userEvent from '@testing-library/user-event'
import DetailsModal from 'components/DetailsModal'
import { TImageExtensions } from 'types'
import { useGetCharacterComics } from 'hooks/endpoints'

jest.mock('hooks/endpoints', () => ({
  useGetCharacterComics: jest.fn(() => ({ data: [], isLoading: false })),
}))

const mockDetailsItem = {
  id: 1,
  name: 'Test Character',
  thumbnail: {
    path: 'test/path',
    extension: 'jpg' as TImageExtensions,
  },
  description: 'Test description',
}

jest.mock('components/ComicsList', () => () => 'comics-list')
beforeEach(() => {
  (useGetCharacterComics as jest.Mock).mockReturnValue({
    data: [],
    isLoading: false,
    isFetching: false,
    refetch: jest.fn(),
  })
})

describe('DetailsModal', () => {
  test('renders DetailsModal correctly', () => {
    const onCloseModalMock = jest.fn()
    const openModal = true

    render(
      <DetailsModal
        detailsItem={mockDetailsItem}
        onCloseModal={onCloseModalMock}
        openModal={openModal}
      />
    )
    expect(screen.getByTestId('details-modal-test-id')).toBeInTheDocument()
    expect(screen.getByText('Details Of Test Character')).toBeInTheDocument()
    expect(screen.getByText('Test Character')).toBeInTheDocument()
    expect(screen.getByText('Test description')).toBeInTheDocument()
    expect(screen.getByText('Comics Of Test Character')).toBeInTheDocument()
    expect(screen.getByText('comics-list')).toBeInTheDocument()
    expect(screen.getByText('Close')).toBeInTheDocument()
  })

  test('calls onCloseModal when "Close" button is clicked', () => {
    const onCloseModalMock = jest.fn()
    const openModal = true

    render(
      <DetailsModal
        detailsItem={mockDetailsItem}
        onCloseModal={onCloseModalMock}
        openModal={openModal}
      />
    )
    const closeButton = screen.getByText('Close')
    closeButton.click()

    expect(onCloseModalMock).toHaveBeenCalled()
  })

  test('matches snapshot', () => {
    const onCloseModalMock = jest.fn()
    const openModal = true

    const { asFragment } = render(
      <DetailsModal
        detailsItem={mockDetailsItem}
        onCloseModal={onCloseModalMock}
        openModal={openModal}
      />
    )
    expect(asFragment()).toMatchSnapshot() // Snapshot test
  })
})
