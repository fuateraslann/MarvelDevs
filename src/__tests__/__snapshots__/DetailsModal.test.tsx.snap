// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsModal matches snapshot 1`] = `<DocumentFragment />`;
